MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java

```diff
@@ -0,0 +1,243 @@
+                GLES20.GL_TEXTURE_MIN_FILTER,
+                GLES20.GL_NEAREST
+        );
+        GLES20.glTexParameterf(
+                GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
+                GLES20.GL_TEXTURE_MAG_FILTER,
+                GLES20.GL_LINEAR
+        );
+        GLES20.glTexParameteri(
+                GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
+                GLES20.GL_TEXTURE_WRAP_S,
+                GLES20.GL_CLAMP_TO_EDGE
+        );
+        GLES20.glTexParameteri(
+                GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
+                GLES20.GL_TEXTURE_WRAP_T,
+                GLES20.GL_CLAMP_TO_EDGE
+        );
+    }
+
+    private static int linkProgram(int vertexShader, int fragmentShader) {
+        Log.d(TAG, "Linking program");
+
+        // Create the program
+        final int program = GLES20.glCreateProgram();
+        if (program == 0) {
+            Log.w(TAG, "Could not create new program");
+            return 0;
+        }
+
+        // Attach shaders to program and link them together
+        GLES20.glAttachShader(program, vertexShader);
+        GLES20.glAttachShader(program, fragmentShader);
+        GLES20.glLinkProgram(program);
+
+        // Verify linking status
+        final int[] linkStatus = new int[1];
+        GLES20.glGetProgramiv(program, GLES20.GL_LINK_STATUS, linkStatus, /* offset = */ 0);
+        if (linkStatus[0] == 0) {
+            // Delete the program if linking failed
+            GLES20.glDeleteProgram(program);
+            Log.w(TAG, "Failed to link program");
+            return 0;
+        }
+
+        return program;
+    }
+
+    private int compileShader(int type, String shaderSrc) {
+        Log.d(TAG, "Compiling shader: " + type);
+
+        // Create new shader object
+        final int shader = GLES20.glCreateShader(type);
+        if (shader == 0) {
+            Log.w(TAG, "Could not create new shader");
+            return 0;
+        }
+
+        // Pass in the shader source
+        GLES20.glShaderSource(shader, shaderSrc);
+
+        // Compile the shader
+        GLES20.glCompileShader(shader);
+
+        // Verify the compile status
+        final int[] compiled = new int[1];
+        GLES20.glGetShaderiv(shader, GLES20.GL_COMPILE_STATUS, compiled, /* offset = */ 0);
+        if (compiled[0] == 0) {
+            // Delete the shader if compilation failed
+            GLES20.glDeleteShader(shader);
+            Log.w(TAG, "Shader failed to compile");
+            return 0;
+        }
+
+        return shader;
+    }
+
+    public void drawFrame(SurfaceTexture surfaceTexture) {
+        Log.d(TAG, "Drawing frame");
+
+        // Set the transform matrix of the surface texture
+        surfaceTexture.getTransformMatrix(uMatrix);
+
+        GLES20.glClearColor(/* red = */ 0.0f, /* green = */ 0.0f, /* blue = */ 0.0f, /* alpha = */ 1.0f);
+        GLES20.glClear(/* mask = */ GLES20.GL_COLOR_BUFFER_BIT);
+
+        GLES20.glUseProgram(program);
+
+        // Process vertices
```

Comment:
useProgram ; glVertexAttribPointer ; glEnableVertexAttribArray ; glActiveTexture ; glBindTexture ; glUniform1i don't need to be executed per frame. OpenGL is stateful and will remember these operations.

Reply:
Changes: moved to onSurfaceCreated(), after linking the program.